packages/cozy-client/src/Schema.js

```diff
@@ -52,13 +63,27 @@ const normalizeDoctypeSchema = doctypeSchema => {
  */
 class Schema {
   constructor(schemaDefinition = {}, client = null) {
+    this.add(schemaDefinition)
+    this.client = client
+  }
+
+  add(schemaDefinition = {}) {
     const values = mapValues(schemaDefinition, (obj, name) => ({
       name,
       ...normalizeDoctypeSchema(obj)
     }))
-    this.client = client
-    this.byName = keyBy(values, x => x.name)
-    this.byDoctype = keyBy(values, x => x.doctype)
+
+    try {
+      this.byName = tryMerge(this.byName, keyBy(values, x => x.name))
+    } catch (error) {
+      throw new Error(`Cannot merge Schema by name (${error.message})`)
+    }
+
+    try {
+      this.byDoctype = tryMerge(this.byDoctype, keyBy(values, x => x.doctype))
+    } catch (error) {
+      throw new Error(`Cannot merge Schema by doctype (${error.message})`)
+    }
```

Comment:
I think those errors could be improved and it should be done before assigning this.byName or this.byDoctype since if you have an error when assigning this.byDoctype, the assignement to this.byName will have succeeded.

I think the check phase and assignment phase should be separated.

Reply:
You're right, that was how my first version was working but then I refactored to this one without paying attention to this point.

Let's go back !

Reply:
I updated the code with better check.

By the way, `this.byName` was never used so I just replaced it by an array of names.

Reply:
> By the way, this.byName was never used

I think we can go further and remove completely schemaNames. Less internal state is good, especially here where schemaNames can be easily computed with `const schemaNames = Object.values(this.byDoctype).map(x => x.name)`

Reply:
You're right, that was my first move to totally remove it but then I get it back to check for the existing names and avoid duplicates.

Your solution is perfect, give me one minute to update the code.